File: lesson_04/kerryferguson/PrimeCheckerTest.java
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import static org.junit.jupiter.api.Assertions.*;

import java.util.Arrays;
import java.util.List;

/**
 * Test suite for PrimeChecker class
 * 
 * Tests the isPrime method and findPrimesInRange method with various edge cases,
 * known primes, composite numbers, and boundary conditions.
 */
public class PrimeCheckerTest {

    @Nested
    @DisplayName("isPrime() method tests")
    class IsPrimeTests {

        @Test
        @DisplayName("Should return false for numbers less than 2")
        void testNumbersLessThanTwo() {
            assertFalse(PrimeChecker.isPrime(-5), "Negative numbers should not be prime");
            assertFalse(PrimeChecker.isPrime(-1), "-1 should not be prime");
            assertFalse(PrimeChecker.isPrime(0), "0 should not be prime");
            assertFalse(PrimeChecker.isPrime(1), "1 should not be prime");
        }

        @Test
        @DisplayName("Should return true for the number 2")
        void testNumberTwo() {
            assertTrue(PrimeChecker.isPrime(2), "2 should be prime");
        }

        @Test
        @DisplayName("Should return false for even numbers greater than 2")
        void testEvenNumbers() {
            assertFalse(PrimeChecker.isPrime(4), "4 should not be prime");
            assertFalse(PrimeChecker.isPrime(6), "6 should not be prime");
            assertFalse(PrimeChecker.isPrime(8), "8 should not be prime");
            assertFalse(PrimeChecker.isPrime(100), "100 should not be prime");
            assertFalse(PrimeChecker.isPrime(1000), "1000 should not be prime");
        }

        @Test
        @DisplayName("Should return true for small known prime numbers")
        void testSmallPrimes() {
            int[] smallPrimes = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47};
            
            for (int prime : smallPrimes) {
                assertTrue(PrimeChecker.isPrime(prime), prime + " should be prime");
            }
        }

        @Test
        @DisplayName("Should return false for small known composite numbers")
        void testSmallComposites() {
            int[] composites = {4, 6, 8, 9, 10, 12, 14, 15, 16, 18, 20, 21, 22, 24, 25, 26, 27, 28};
            
            for (int composite : composites) {
                assertFalse(PrimeChecker.isPrime(composite), composite + " should not be prime");
            }
        }

        @Test
        @DisplayName("Should return true for larger known prime numbers")
        void testLargerPrimes() {
            int[] largerPrimes = {53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113};
            
            for (int prime : largerPrimes) {
                assertTrue(PrimeChecker.isPrime(prime), prime + " should be prime");
            }
        }

        @Test
        @DisplayName("Should return false for perfect squares")
        void testPerfectSquares() {
            assertFalse(PrimeChecker.isPrime(9), "9 (3²) should not be prime");
            assertFalse(PrimeChecker.isPrime(25), "25 (5²) should not be prime");
            assertFalse(PrimeChecker.isPrime(49), "49 (7²) should not be prime");
            assertFalse(PrimeChecker.isPrime(121), "121 (11²) should not be prime");
            assertFalse(PrimeChecker.isPrime(169), "169 (13²) should not be prime");
        }

        @Test
        @DisplayName("Should handle edge cases correctly")
        void testEdgeCases() {
            // Test some tricky numbers that might fool naive algorithms
            assertFalse(PrimeChecker.isPrime(91), "91 (7 × 13) should not be prime");
            assertFalse(PrimeChecker.isPrime(93), "93 (3 × 31) should not be prime");
            assertFalse(PrimeChecker.isPrime(95), "95 (5 × 19) should not be prime");
            
            // Test some larger primes
            assertTrue(PrimeChecker.isPrime(127), "127 should be prime");
            assertTrue(PrimeChecker.isPrime(131), "131 should be prime");
        }
    }

    @Nested
    @DisplayName("findPrimesInRange() method tests")
    class FindPrimesInRangeTests {

        @Test
        @DisplayName("Should return empty list for invalid ranges")
        void testInvalidRanges() {
            List<Integer> result1 = PrimeChecker.findPrimesInRange(5, 2);
            assertTrue(result1.isEmpty(), "Should return empty list when start > end");
            
            List<Integer> result2 = PrimeChecker.findPrimesInRange(-5, 1);
            assertTrue(result2.isEmpty(), "Should return empty list for range with no primes");
        }

        @Test
        @DisplayName("Should find primes in range 1-10")
        void testRange1To10() {
            List<Integer> expected = Arrays.asList(2, 3, 5, 7);
            List<Integer> actual = PrimeChecker.findPrimesInRange(1, 10);
            assertEquals(expected, actual, "Should find primes 2, 3, 5, 7 in range 1-10");
        }

        @Test
        @DisplayName("Should find primes in range 10-30")
        void testRange10To30() {
            List<Integer> expected = Arrays.asList(11, 13, 17, 19, 23, 29);
            List<Integer> actual = PrimeChecker.findPrimesInRange(10, 30);
            assertEquals(expected, actual, "Should find correct primes in range 10-30");
        }

        @Test
        @DisplayName("Should find primes in range with single prime")
        void testSinglePrimeRange() {
            List<Integer> expected = Arrays.asList(7);
            List<Integer> actual = PrimeChecker.findPrimesInRange(7, 7);
            assertEquals(expected, actual, "Should find single prime 7 in range 7-7");
        }

        @Test
        @DisplayName("Should return empty list for range with no primes")
        void testRangeWithNoPrimes() {
            List<Integer> actual = PrimeChecker.findPrimesInRange(24, 28);
            assertTrue(actual.isEmpty(), "Should return empty list for range 24-28 (no primes)");
        }

        @Test
        @DisplayName("Should find all primes up to 50")
        void testPrimesUpTo50() {
            List<Integer> expected = Arrays.asList(
                2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47
            );
            List<Integer> actual = PrimeChecker.findPrimesInRange(1, 50);
            assertEquals(expected, actual, "Should find all 15 primes up to 50");
            assertEquals(15, actual.size(), "Should find exactly 15 primes up to 50");
        }
    }

    @Nested
    @DisplayName("Performance and boundary tests")
    class PerformanceTests {

        @Test
        @DisplayName("Should handle larger numbers efficiently")
        void testLargerNumbers() {
            // Test some larger primes - these should complete quickly
            assertTrue(PrimeChecker.isPrime(997), "997 should be prime");
            assertTrue(PrimeChecker.isPrime(1009), "1009 should be prime");
            
            // Test some larger composites
            assertFalse(PrimeChecker.isPrime(1000), "1000 should not be prime");
            assertFalse(PrimeChecker.isPrime(1001), "1001 should not be prime");
        }

        @Test
        @DisplayName("Should find correct number of primes in larger range")
        void testLargerRange() {
            List<Integer> primes = PrimeChecker.findPrimesInRange(1, 100);
            assertEquals(25, primes.size(), "Should find exactly 25 primes between 1 and 100");
            
            // Verify the last few primes in this range
            assertTrue(primes.contains(89), "Should contain 89");
            assertTrue(primes.contains(97), "Should contain 97");
            assertFalse(primes.contains(91), "Should not contain 91");
            assertFalse(primes.contains(93), "Should not contain 93");
        }
    }
}

File: lesson_04/kerryferguson/prime_checker.test.js
/**
 * Test suite for prime_checker.js
 * 
 * This module contains comprehensive unit tests for the JavaScript prime number checker
 * implementation, including edge cases, boundary conditions, and performance tests.
 */

const { isPrime, findPrimesInRange, main } = require('./prime_checker');

describe('isPrime function', () => {
    describe('numbers less than 2', () => {
        test('should return false for negative numbers', () => {
            expect(isPrime(-5)).toBe(false);
            expect(isPrime(-1)).toBe(false);
        });
        
        test('should return false for 0 and 1', () => {
            expect(isPrime(0)).toBe(false);
            expect(isPrime(1)).toBe(false);
        });
    });
    
    describe('number 2', () => {
        test('should return true for 2', () => {
            expect(isPrime(2)).toBe(true);
        });
    });
    
    describe('even numbers greater than 2', () => {
        test('should return false for even numbers', () => {
            const evenNumbers = [4, 6, 8, 10, 12, 100, 1000];
            evenNumbers.forEach(num => {
                expect(isPrime(num)).toBe(false);
            });
        });
    });
    
    describe('small known prime numbers', () => {
        test('should return true for small primes', () => {
            const smallPrimes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47];
            smallPrimes.forEach(prime => {
                expect(isPrime(prime)).toBe(true);
            });
        });
    });
    
    describe('small known composite numbers', () => {
        test('should return false for small composites', () => {
            const composites = [4, 6, 8, 9, 10, 12, 14, 15, 16, 18, 20, 21, 22, 24, 25, 26, 27, 28];
            composites.forEach(composite => {
                expect(isPrime(composite)).toBe(false);
            });
        });
    });
    
    describe('larger known prime numbers', () => {
        test('should return true for larger primes', () => {
            const largerPrimes = [53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113];
            largerPrimes.forEach(prime => {
                expect(isPrime(prime)).toBe(true);
            });
        });
    });
    
    describe('perfect squares', () => {
        test('should return false for perfect squares', () => {
            const perfectSquares = [9, 25, 49, 121, 169, 289]; // 3², 5², 7², 11², 13², 17²
            perfectSquares.forEach(square => {
                expect(isPrime(square)).toBe(false);
            });
        });
    });
    
    describe('edge cases', () => {
        test('should handle tricky composite numbers', () => {
            // These are composite numbers that might be mistakenly identified as prime
            const trickyComposites = [91, 93, 95, 121, 143, 169, 187, 209];
            trickyComposites.forEach(num => {
                expect(isPrime(num)).toBe(false);
            });
        });
        
        test('should handle larger primes correctly', () => {
            const largerPrimes = [127, 131, 137, 139, 149];
            largerPrimes.forEach(prime => {
                expect(isPrime(prime)).toBe(true);
            });
        });
    });
});

describe('findPrimesInRange function', () => {
    describe('invalid ranges', () => {
        test('should return empty array when start > end', () => {
            const result = findPrimesInRange(10, 5);
            expect(result).toEqual([]);
        });
        
        test('should return empty array for range with no primes', () => {
            const result = findPrimesInRange(-5, 1);
            expect(result).toEqual([]);
        });
    });
    
    describe('valid ranges', () => {
        test('should find primes in range 1-10', () => {
            const expected = [2, 3, 5, 7];
            const actual = findPrimesInRange(1, 10);
            expect(actual).toEqual(expected);
        });
        
        test('should find primes in range 10-30', () => {
            const expected = [11, 13, 17, 19, 23, 29];
            const actual = findPrimesInRange(10, 30);
            expect(actual).toEqual(expected);
        });
        
        test('should handle single number ranges', () => {
            // Single prime
            expect(findPrimesInRange(7, 7)).toEqual([7]);
            // Single composite
            expect(findPrimesInRange(8, 8)).toEqual([]);
        });
        
        test('should return empty array for range with no primes', () => {
            const result = findPrimesInRange(24, 28);
            expect(result).toEqual([]);
        });
        
        test('should find all primes up to 50', () => {
            const expected = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47];
            const actual = findPrimesInRange(1, 50);
            expect(actual).toEqual(expected);
            expect(actual).toHaveLength(15);
        });
        
        test('should find correct number of primes up to 100', () => {
            const primes = findPrimesInRange(1, 100);
            expect(primes).toHaveLength(25);
            expect(primes).toContain(89);
            expect(primes).toContain(97);
            expect(primes).not.toContain(91);
            expect(primes).not.toContain(93);
        });
    });
});

describe('main function', () => {
    let consoleSpy;
    
    beforeEach(() => {
        consoleSpy = jest.spyOn(console, 'log').mockImplementation(() => {});
    });
    
    afterEach(() => {
        consoleSpy.mockRestore();
    });
    
    test('should output expected content', () => {
        main();
        
        // Check that key elements are in the output
        expect(consoleSpy).toHaveBeenCalledWith(
            expect.stringContaining("Prime Number Checker - JavaScript Implementation")
        );
        expect(consoleSpy).toHaveBeenCalledWith(
            expect.stringContaining("Testing individual numbers")
        );
    });
    
    test('should test individual numbers correctly', () => {
        main();
        
        // Check specific test outputs
        expect(consoleSpy).toHaveBeenCalledWith(
            expect.stringMatching(/17.*Prime/)
        );
        expect(consoleSpy).toHaveBeenCalledWith(
            expect.stringMatching(/25.*Not Prime/)
        );
    });
});

describe('performance tests', () => {
    test('should handle larger primes efficiently', () => {
        const largerPrimes = [997, 1009, 1013, 1019, 1021];
        largerPrimes.forEach(prime => {
            expect(isPrime(prime)).toBe(true);
        });
    });
    
    test('should handle larger composite numbers', () => {
        const largerComposites = [1000, 1001, 1002, 1004, 1006];
        largerComposites.forEach(composite => {
            expect(isPrime(composite)).toBe(false);
        });
    });
    
    test('should efficiently count primes in ranges', () => {
        const primesTo100 = findPrimesInRange(1, 100);
        expect(primesTo100).toHaveLength(25);
        
        const primesTo200 = findPrimesInRange(1, 200);
        expect(primesTo200).toHaveLength(46);
    });
    
    test('should complete large prime checks quickly', () => {
        const start = performance.now();
        const result = isPrime(982451653); // Large prime
        const end = performance.now();
        
        expect(result).toBe(true);
        expect(end - start).toBeLessThan(100); // Should complete in less than 100ms
    });
});

describe('algorithm consistency', () => {
    test('should produce same results as mathematical definition', () => {
        // Test against known sequence of first 25 primes
        const first25Primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97];
        
        first25Primes.forEach(prime => {
            expect(isPrime(prime)).toBe(true);
        });
        
        // Test that no numbers between these primes are incorrectly identified as prime
        for (let i = 2; i <= 97; i++) {
            if (!first25Primes.includes(i)) {
                expect(isPrime(i)).toBe(false);
            }
        }
    });
});
